File: src/TheCarl.Application/Sync/SyncContracts.cs
using TheCarl.Domain;

namespace TheCarl.Application.Sync;

/// <summary>Outcome of one transaction inside a batch.</summary>
public enum SyncItemStatus
{
    /// <summary>Posted to the ledger for the first time.</summary>
    Accepted,

    /// <summary>
    /// Already recorded. A success, not an error: it means an earlier attempt landed.
    /// Clients must mark the row synced and stop retrying.
    /// </summary>
    Duplicate,

    /// <summary>Permanently invalid. Retrying will never succeed.</summary>
    Rejected,

    /// <summary>Server state disagrees with the submission. Needs a human.</summary>
    Conflict
}

/// <summary>
/// Stable, machine-readable reason codes.
/// </summary>
/// <remarks>
/// These are part of the public API contract: the Android client branches on them. Values
/// are never renamed or repurposed — a new situation gets a new code.
/// </remarks>
public static class SyncReasonCodes
{
    // Duplicate
    public const string DuplicateClientTransactionId = "DUPLICATE_CLIENT_TRANSACTION_ID";
    public const string DuplicateEvidenceFingerprint = "DUPLICATE_EVIDENCE_FINGERPRINT";

    // Rejected — payload
    public const string InvalidClientTransactionId = "INVALID_CLIENT_TRANSACTION_ID";
    public const string InvalidAmount = "INVALID_AMOUNT";
    public const string InvalidTransactionType = "INVALID_TRANSACTION_TYPE";
    public const string InvalidProvider = "INVALID_PROVIDER";
    public const string InvalidEvidenceFingerprint = "INVALID_EVIDENCE_FINGERPRINT";
    public const string InvalidSourceType = "INVALID_SOURCE_TYPE";
    public const string MissingParserVersion = "MISSING_PARSER_VERSION";

    // Rejected — temporal
    public const string TimestampInFuture = "TIMESTAMP_IN_FUTURE";
    public const string TimestampTooOld = "TIMESTAMP_TOO_OLD";

    // Rejected — accounting rules
    public const string UnknownTypeRequiresReview = "UNKNOWN_TYPE_REQUIRES_REVIEW";
    public const string ReversalRequiresOriginal = "REVERSAL_REQUIRES_ORIGINAL";
    public const string ReversalOriginalNotFound = "REVERSAL_ORIGINAL_NOT_FOUND";
    public const string ReversalOriginalNotEligible = "REVERSAL_ORIGINAL_NOT_ELIGIBLE";
    public const string AdjustmentRequiresExplicitDeltas = "ADJUSTMENT_REQUIRES_EXPLICIT_DELTAS";
    public const string AdjustmentRequiresReason = "ADJUSTMENT_REQUIRES_REASON";

    // Rejected — authorization and relationships
    public const string BranchNotInTenant = "BRANCH_NOT_IN_TENANT";
    public const string DeviceNotInTenant = "DEVICE_NOT_IN_TENANT";
    public const string DeviceRevoked = "DEVICE_REVOKED";
    public const string SessionNotInTenant = "SESSION_NOT_IN_TENANT";
    public const string SessionBranchMismatch = "SESSION_BRANCH_MISMATCH";
    public const string SessionNotOpenAtEventTime = "SESSION_NOT_OPEN_AT_EVENT_TIME";

    // Conflict
    public const string ClientIdReusedWithDifferentPayload = "CLIENT_ID_REUSED_WITH_DIFFERENT_PAYLOAD";

    /// <summary>
    /// The original transaction already has an effective reversal. Reversing twice would
    /// create money that never existed, so the second attempt is refused and raised as a
    /// conflict for a person to look at.
    /// </summary>
    public const string AlreadyReversed = "ALREADY_REVERSED";

    // Transient — retryable with backoff
    public const string DatabaseUnavailable = "DATABASE_UNAVAILABLE";
    public const string Timeout = "TIMEOUT";
    public const string TemporaryServerError = "TEMPORARY_SERVER_ERROR";
}

/// <summary>
/// One transaction in a sync batch.
/// </summary>
/// <remarks>
/// There is no <c>OrganizationId</c>: the tenant comes from the access token. Every field
/// here is untrusted client input and is validated server-side, regardless of any validation
/// the Android client performed for its own UX.
/// </remarks>
public sealed record SyncTransactionItem(
    string ClientTransactionId,
    TransactionType TransactionType,
    decimal Amount,
    string Provider,
    DateTimeOffset TransactionTimestamp,
    DateTimeOffset? DeviceReceivedAt = null,
    Guid? BranchId = null,
    Guid? DeviceId = null,
    Guid? SessionId = null,
    string? Currency = null,
    string? CustomerPhone = null,
    string? TransactionReference = null,
    string? EvidenceFingerprint = null,
    string? ParserVersion = null,
    EvidenceSourceType SourceType = EvidenceSourceType.ManualEntry,
    Guid? ReversesTransactionId = null,
    decimal? AdjustmentCashDelta = null,
    decimal? AdjustmentFloatDelta = null,
    string? CorrectionReason = null,
    string? Notes = null);

/// <summary>A batch of transactions captured offline.</summary>
public sealed record SyncTransactionsRequest(IReadOnlyList<SyncTransactionItem> Transactions);

/// <summary>
/// Per-item outcome. Deliberately narrow: it carries what the client needs to advance its
/// outbox and nothing more. No evidence bodies, no internal identifiers.
/// </summary>
public sealed record SyncTransactionResult(
    string ClientTransactionId,
    SyncItemStatus Status,
    Guid? TransactionId,
    string? ReasonCode,
    string? Message)
{
    /// <summary>
    /// Stable category for this outcome. Clients branch on this and <see cref="ReasonCode"/>,
    /// never on <see cref="Message"/>, which is written for people and may change.
    /// </summary>
    public SyncErrorCategory Category => SyncErrorClassification.CategoryFor(ReasonCode);

    /// <summary>Whether retrying this item could ever succeed.</summary>
    public bool IsRetryable => SyncErrorClassification.IsRetryable(Category);

    /// <summary>Conflict record raised for this item, when one was created.</summary>
    public Guid? ConflictId { get; init; }
}

/// <summary>Batch outcome with per-item results and counts.</summary>
public sealed record SyncTransactionsResponse(
    Guid BatchId,
    int Submitted,
    int Accepted,
    int Duplicate,
    int Rejected,
    int Conflict,
    DateTimeOffset ServerReceivedAtUtc,
    IReadOnlyList<SyncTransactionResult> Results);

/// <summary>
/// Trusted context for a sync batch, resolved server-side from the access token.
/// Nothing here originates from the request body.
/// </summary>
public sealed record SyncCallerContext(
    Guid OrganizationId,
    Guid UserId,
    Guid? CallerBranchId,
    bool HasOrganizationWideScope,
    string CorrelationId);

public interface ISyncTransactionService
{
    Task<SyncTransactionsResponse> SynchronizeAsync(
        SyncTransactionsRequest request,
        SyncCallerContext caller,
        CancellationToken cancellationToken = default);
}

File: android/core/domain/src/main/kotlin/app/zazi/core/domain/sync/SyncContracts.kt
package app.zazi.core.domain.sync

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Wire contract for `POST /api/v1/sync/transactions`.
 *
 * **Mirrors the backend exactly.** These types were written against the .NET
 * `SyncTransactionApiItem` / `SyncTransactionsResponse` records, not invented. Field names
 * match the server's camelCase JSON.
 *
 * Note there is deliberately **no `organizationId`**. The server derives tenancy from the
 * access token; a client cannot supply or override it. See docs/ANDROID_API_CONTRACT.md.
 */
@Serializable
data class SyncTransactionRequestItem(
    val clientTransactionId: String,
    val transactionType: Int,
    val amount: String,
    val provider: String,
    val transactionTimestamp: String,
    val deviceReceivedAt: String? = null,
    val branchId: String? = null,
    val deviceId: String? = null,
    val sessionId: String? = null,
    val currency: String? = null,
    val customerPhone: String? = null,
    val transactionReference: String? = null,
    val evidenceFingerprint: String? = null,
    val parserVersion: String? = null,
    val sourceType: Int,
    val reversesTransactionId: String? = null,
    val adjustmentCashDelta: String? = null,
    val adjustmentFloatDelta: String? = null,
    val correctionReason: String? = null,
    val notes: String? = null,
    /**
     * The counterparty's registered name, where the network stated one.
     *
     * <p>Last in the list to match the server's contract, which appends rather than inserts:
     * a parameter added to the middle of a positional record shifts every call site, and the
     * compiler only notices where the types differ.</p>
     */
    val customerName: String? = null
)

@Serializable
data class SyncTransactionsRequest(
    val transactions: List<SyncTransactionRequestItem>
)

@Serializable
data class SyncTransactionResult(
    val clientTransactionId: String,
    val status: String,
    val transactionId: String? = null,
    val reasonCode: String? = null,
    val message: String? = null,
    val category: String? = null,
    val isRetryable: Boolean = false,
    val conflictId: String? = null
)

@Serializable
data class SyncTransactionsResponse(
    val batchId: String,
    val submitted: Int,
    val accepted: Int,
    val duplicate: Int,
    val rejected: Int,
    val conflict: Int,
    @SerialName("serverReceivedAtUtc") val serverReceivedAtUtc: String,
    val results: List<SyncTransactionResult>
)

/** Server-side per-item outcomes, as returned in [SyncTransactionResult.status]. */
object SyncItemStatus {
    const val ACCEPTED = "Accepted"

    /**
     * Already recorded. **A success, not an error.** It means an earlier attempt landed —
     * typically one whose response was lost — and the client must mark the row synced using
     * the returned transaction id rather than retrying.
     */
    const val DUPLICATE = "Duplicate"

    const val REJECTED = "Rejected"
    const val CONFLICT = "Conflict"
}

/** Stable server reason codes the client branches on. */
object SyncReasonCodes {
    const val DUPLICATE_CLIENT_TRANSACTION_ID = "DUPLICATE_CLIENT_TRANSACTION_ID"
    const val DUPLICATE_EVIDENCE_FINGERPRINT = "DUPLICATE_EVIDENCE_FINGERPRINT"
    const val ALREADY_REVERSED = "ALREADY_REVERSED"
    const val DEVICE_REVOKED = "DEVICE_REVOKED"
    const val BRANCH_NOT_IN_TENANT = "BRANCH_NOT_IN_TENANT"
    const val DEVICE_NOT_IN_TENANT = "DEVICE_NOT_IN_TENANT"
}

/**
 * Maps a server result onto the local outbox state.
 *
 * This is the single place the client decides what a response means, so the lost-response
 * rule cannot be implemented differently in two call sites.
 */
object SyncResultInterpreter {

    /**
     * @param attemptsSoFar attempts already made, including the one that produced [result].
     */
    fun nextState(result: SyncTransactionResult, attemptsSoFar: Int): OutboxState = when (result.status) {
        SyncItemStatus.ACCEPTED -> OutboxState.SYNCED

        // The lost-response rule. The server already holds this transaction; retrying would
        // never produce anything different, and treating it as failure is what turns one
        // transaction into a retry storm.
        SyncItemStatus.DUPLICATE -> OutboxState.SYNCED

        SyncItemStatus.CONFLICT -> OutboxState.CONFLICT

        SyncItemStatus.REJECTED ->
            if (result.isRetryable && RetryPolicy.hasBudgetRemaining(attemptsSoFar)) {
                OutboxState.RETRYABLE_FAILURE
            } else {
                // Permanently invalid, or out of budget. Never deleted — an agent's record of
                // work must remain inspectable even when it cannot be posted.
                OutboxState.DEAD_LETTER
            }

        // An unrecognised status means the server moved ahead of this client. Retry rather
        // than dead-letter: guessing "permanent" would discard a possibly-valid transaction.
        else ->
            if (RetryPolicy.hasBudgetRemaining(attemptsSoFar)) OutboxState.RETRYABLE_FAILURE
            else OutboxState.DEAD_LETTER
    }

    /**
     * The server transaction id to record locally, if any. Present for both acceptances and
     * duplicates, which is what lets a client reconcile after losing a response.
     */
    fun serverTransactionId(result: SyncTransactionResult): String? = result.transactionId
}

/**
 * How a transport-level failure — no response at all — maps onto the outbox.
 *
 * **A timeout is not a failure to record.** The server may have committed. The row returns
 * to a retryable state carrying the same ClientTransactionId, and the server's idempotency
 * constraint resolves it as a duplicate on the next attempt.
 */
object TransportFailurePolicy {

    fun nextState(httpStatus: Int?, attemptsSoFar: Int): OutboxState {
        val budget = RetryPolicy.hasBudgetRemaining(attemptsSoFar)

        return when (httpStatus) {
            // No response: connection reset, timeout, airplane mode. Always retryable.
            null -> if (budget) OutboxState.RETRYABLE_FAILURE else OutboxState.DEAD_LETTER

            // Authentication and authorization: retrying the batch will not help until the
            // session is repaired, but the transaction itself is still valid work.
            401, 403 -> OutboxState.RETRYABLE_FAILURE

            // Rate limited. Always retryable; the worker honours Retry-After.
            429 -> OutboxState.RETRYABLE_FAILURE

            // Server-side faults are transient by definition.
            in 500..599 -> if (budget) OutboxState.RETRYABLE_FAILURE else OutboxState.DEAD_LETTER

            // The batch itself was malformed or too large. Retrying it unchanged cannot help.
            400, 413 -> OutboxState.DEAD_LETTER

            else -> if (budget) OutboxState.RETRYABLE_FAILURE else OutboxState.DEAD_LETTER
        }
    }

    /** Whether a 401 should trigger a token refresh before the next attempt. */
    fun requiresTokenRefresh(httpStatus: Int?): Boolean = httpStatus == 401

    /**
     * Whether the local authentication state should be cleared outright. A 403 after a
     * successful refresh means the device or session was revoked server-side, and continuing
     * to retry would be pointless.
     */
    fun indicatesRevocation(httpStatus: Int?, reasonCode: String?): Boolean =
        httpStatus == 403 || reasonCode == SyncReasonCodes.DEVICE_REVOKED
}
